pkg/lock/semaphored_mutex.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lock
+
+import (
+	"context"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// SemaphoredMutex is a semaphored mutex that provodes a RWLocker interface.
+type SemaphoredMutex struct {
+	readers int64
+	*semaphore.Weighted
```

Comment:
We should hide the semaphore to prevent `Acquire()` and `Release()` calls being made directly.